mirgecom/profiling.py

```diff
@@ -0,0 +1,202 @@
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        for t in self.profile_events:
+            program = t.program
+            res = self._get_kernel_stats(program, t.kwargs)
+            res.time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            self.profile_results.setdefault(program, []).append(res)
+
+        self.profile_events = []
+
+    def table_profiling_data(self) -> pytools.Table:
+        """Returns a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls", "Time_min (s)", "Time_avg (s)",
+            "Time_max (s)", "GFlops_min", "GFlops_avg", "GFlops_max", "GMemAcc_min",
+            "GMemAcc_avg", "GMemAcc_max", "Bandwidth (GByte/s)",
+            "Footprint (GByte)"])
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v.time / 1e9 for v in value]
+            flops = [v.flops / 1e9 for v in value]
+            bytes_accessed = [v.bytes_accessed / 1e9 for v in value]
+            footprint_bytes = [v.footprint_bytes / 1e9
+                if v.footprint_bytes is not None else 0 for v in value]
+
+            tbl.add_row([key.name, num_values, min(times),
+                mean(times), max(times), min(flops), mean(flops),
+                max(flops), min(bytes_accessed), mean(bytes_accessed),
+                max(bytes_accessed), round(mean(bytes_accessed)/mean(times), 3),
+                mean(footprint_bytes)])
+
+        return tbl
+
+    def _get_kernel_stats(self, program, kwargs: dict) -> ProfileResult:
+        # We need a tuple to index the cache
+        args_tuple = tuple(
+            (key, value.shape) if hasattr(value, "shape") else (key, value)
+            for key, value in kwargs.items())
+
+        if (program not in self.kernel_stats
+                or args_tuple not in self.kernel_stats[program]):
+            kwargs = dict(kwargs)
+            executor = program.target.get_kernel_executor(program, self.queue)
+            info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+            kernel = executor.get_typed_and_scheduled_kernel(
+                executor.arg_to_dtype_set(kwargs))
+
+            idi = info.implemented_data_info
+
+            types = {k: v for k, v in kwargs.items()
+                if hasattr(v, "dtype") and not v.dtype == object}
+
+            param_dict = {**kwargs}
+            param_dict.update({k: None for k, value in kernel.arg_dict.items()
+                if k not in param_dict})
+
+            param_dict.update(
+                {d.name: None for d in idi if d.name not in param_dict})
+
+            # Generate the wrapper code
+            wrapper = executor.get_wrapper_generator()
+
+            gen = PythonFunctionGenerator("my_gen_args", ["param_dict"])
```

Comment:
Note that this code runs in a namespace that is under user (and hence not under your) control. That means all identifiers you introduce must be prefixed somehow.

Reply:
renamed to `_my_gen_args_profiling`.

Reply:
`_` doesn't suffice as a prefix. `_mcom_`? Also, in the previous version of the code `param_dict` was a name to which this would apply.

Reply:
Changed prefix to `_mcom_`.